2023 05 17 add liveness methods

andrewbaptist opened the pull request.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/103677)
<!-- Reviewable:end -->

Commit:
storepool: Simplify liveness func

Remove unnnecessary methods from the liveness func.

Epic: none

Release note: None

Commit:
liveness: use simpler method for admin liveness

Change to use the liveness method that determines status based on
NodeStatus rather than on liveness directly.

Epic: none

Release note: None

Commit:
liveness: remove GetNodeCount

GetNodeCount is very specific to the StorePool and it is a confusing
method in general to have on the interface. It is much better to
directly compute this based on the cached values when needed.

Epic: none

Release note: None

Commit:
liveness: Move TestNodeLivenessLivenessStatus to liveness

Previously this method was in the storepool package, however it makes
more sense to have this in the liveness package as it is testing the
functionality there.

Epic: none

Release note: None

Commit:
convert SQL and upgrade to NodeStatus

Liveness is not intended to be used directly. This converts the usages
in upgrade and the internal crdb tables that expose liveness to instead
use the NodeStatus object. This makes it easier to change the IsLive
condition.

Epic: none

Release note: None

Commit:
liveness: remove GetLivenessesFromKV

replace GetLivenessesFromKV with GetNodeStatusFromKV in all places and
remove the unnecessary method.

Epic: none

Release note: None

Commit:
liveness: remove some cruft

Remove some unnecessary functions and clean up. Prepare to allow
stopping updating the expiration timestamp in the liveness record.

Epic: none

Release note: None

Commit:
server: remove unused LivenessStatus field

This field was returned in the response to a DecommissionPreCheck, but
the value was never read. In a seperate PR we are attempting to remove
the usage of this field, so removing it here helps.

Epic: none

Release note: None

Commit:
server: reduce usage of NodeLivenesssStatus

NodeLivenessStatus mixes too many concepts together and makes it hard
for the recipient to know all the necessary information. It is possible
for a node to be in various combinations of decomisssioning, draining,
unavailable or dead, and the NodeLivenessStatus doesn't allow that
differentiation. The field can't be removed completely yet because it
is in the public API, but future PRs will remove its usage.

Epic: none

Release note: None

Commit:
liveness: remove the IsLive function from liveness

The cleaner way to do this is to load the node from the cache and check
its liveness directly.

Epic: none

Release note: None

Commit:
kvserver: change replica range lease to use NodeVitality

Epic: none

Release note: None

Commit:
liveness: change IsLiveMap to return NodeVitality

This changes the IsLiveMap function to be consistent with the other
methods in Liveness and return NodeVitality instead.

Epic: none

Release note: None

Commit:
liveness: remove unnecessary Liveness interfaces

NodeLiveness wasn't previously split out as an interface. That meant
that each of the places that used it needed to create a subset of the
interfaces that it used. Now that there are three primary methods, and
an NodeLivenessInterface, it is much easier for the code to simply use
the interface directly.

Epic: none

Release note: None

Commit:
liveness: Assume if epoch is 0 that the node is alive

WIP: This makes some tests pass and others fails. Need to figure out the
right approach here.

Epic: none

Release note: None
